File: src/p0016/cpp/solution.cpp
class Solution {
public:
    int threeSumClosest(vector<int> &nums, int target) {
        int minSum = 0, minDelta = INT_MAX;
        sort(nums.begin(), nums.end());
        for (size_t i = 0; i < nums.size(); i++) {
            for (size_t j = i + 1, k = nums.size() - 1; j < k;) {
                int sum = nums[i] + nums[j] + nums[k];
                if (sum < target) {
                    j++;
                } else if (sum > target) {
                    k--;
                } else {
                    return sum;
                }
                int delta = abs(sum - target);
                if (minDelta > delta) {
                    minSum = sum;
                    minDelta = delta;
                }
            }
        }
        return minSum;
    }
};

File: src/p0016/rust/solution.rs
impl Solution {
    pub fn three_sum_closest(nums: Vec<i32>, target: i32) -> i32 {
        let (mut min_sum, mut min_delta) = (0, std::i32::MAX);
        let mut nums = nums;
        nums.sort();
        for i in 0..nums.len() {
            let (mut j, mut k) = (i + 1, nums.len() - 1);
            while j < k {
                let sum = nums[i] + nums[j] + nums[k];
                if sum < target {
                    j += 1;
                } else if sum > target {
                    k -= 1;
                } else {
                    return sum;
                }
                let delta = i32::abs(sum - target);
                if min_delta > delta {
                    min_sum = sum;
                    min_delta = delta;
                }
            }
        }
        return min_sum;
    }
}
